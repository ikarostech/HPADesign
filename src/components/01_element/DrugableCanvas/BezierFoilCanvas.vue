<template>
  <canvas
    ref="drawCanvas"
    :width="width"
    :height="height"
    v-on:mousedown="mousedown"
    v-on:mouseup="mouseup"
    v-on:mousemove="mousemove"
  />
</template>

<script lang="ts">
import assert from "assert";
import { LinearInterpolite } from "@/model/math/LinearInterpolite/LinearInterpolite";
import { Vector2 } from "three/src/math/Vector2";
import Vue, { PropType } from "vue";
import { BezierControls } from "./BezierControls";
import { CanvasPoint } from "./CanvasPoint";

//import { Vector2 } from 'three/src/math/Vector2';

export function buildBezierControls(
  controls: BezierControls,
  linspaceSize: number
): Vector2[] {
  const bezierCurve: Vector2[] = [];

  //TODO BezierPointの仕様変更
  for (let i = 0; i < controls.points.length - 1; i += 2) {
    const p1: Vector2 = controls.points[i].point;
    const c: Vector2 = controls.points[i + 1].point;
    const p2: Vector2 = controls.points[i + 2].point;

    for (let n = 0; n < linspaceSize; n++) {
      const t: number = n / linspaceSize;

      const pos: Vector2 = p1
        .clone()
        .multiplyScalar(Math.pow(1 - t, 2))
        .add(c.clone().multiplyScalar(2 * t * (1 - t)))
        .add(p2.clone().multiplyScalar(Math.pow(t, 2)));

      bezierCurve.push(pos);
    }
  }
  bezierCurve.push(new Vector2(1, 0));
  return bezierCurve;
}

export type select = {
  isTop: boolean;
  index: number;
}
export type DataType = {
  topControls: BezierControls;
  canverControls: BezierControls;
  canvas: any;
  ctx: any;
  width: number;
  height: number;
	isDrag: boolean;
	selected: select | null;
};

export default Vue.extend({
  name: "App",
  props: {
    radius: Number,
  },
  data: function (): DataType {
    const topControls = new BezierControls();
    topControls.points = [
      {
            point: new Vector2(0.0, 0.0),
            visible: false,
          },
          {
            point: new Vector2(0.0, 0.09),
            visible: true,
          },
          {
            point: new Vector2(0.34, 0.09),
            visible: true,
          },
          {
            point: new Vector2(0.6, 0.09),
            visible: true,
          },
          {
            point: new Vector2(1.0, 0.0),
            visible: false,
          },
    ];
    const canverControls = new BezierControls();
    canverControls.points = [
          {
            point: new Vector2(0.0, 0.0),
            visible: false,
          },
          {
            point: new Vector2(0.2, 0.03),
            visible: true,
          },
          {
            point: new Vector2(0.55, 0.03),
            visible: true,
          },
          {
            point: new Vector2(0.65, 0.03),
            visible: true,
          },
          {
            point: new Vector2(1, 0),
            visible: false,
          },
    ];
    return {
      topControls: topControls,
      canverControls: canverControls,
      canvas: null,
      ctx: null,
      width: 600,
      height: 200,
			isDrag: false,
      selected: null,
    };
  },

  mounted: function () {
    this.canvas = this.$refs.drawCanvas;
    this.ctx = this.canvas.getContext("2d");
    this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
    this.draw();
  },
  computed: {
    getShape: function (): Vector2[] {
      //翼型上面
      const topShape: Vector2[] = LinearInterpolite(
        buildBezierControls(this.topControls, 100),
        81
      );
      const canverShape: Vector2[] = LinearInterpolite(
        buildBezierControls(this.canverControls, 100),
        81
      );
      assert(topShape.length == canverShape.length);

      const bottomShape: Vector2[] = [];
      for (let i = 1; i < topShape.length; i++) {
        const element: Vector2 = canverShape[i]
          .multiplyScalar(2)
          .sub(topShape[i]);
        bottomShape.push(element);
      }
      assert(topShape.length == bottomShape.length + 1);

      const shape = topShape.reverse().concat(bottomShape);
      this.$emit("updateShape", shape);
      return shape;
    },
  },

  methods: {
    mousedown: function (event: {target: HTMLCanvasElement; clientX: number; clientY: number }): void {
			this.isDrag = true;
			this.selected = null;

			const rect = event.target.getBoundingClientRect();
      const mousePos = new Vector2(event.clientX - rect.left - 100, -(event.clientY - rect.top - 100)).divideScalar(400);

      const topControlFirstMatch: number = this.topControls.points
        .findIndex(controlPoint => controlPoint.visible === true && controlPoint.point.distanceTo(mousePos) <= this.radius / 400)
      
      if (topControlFirstMatch >= 0) {
        this.selected = {
          isTop: true,
          index: topControlFirstMatch
        };
        return;
      }

      const canverControlFirstMatch: number = this.canverControls.points
        .findIndex(controlPoint => controlPoint.visible === true && controlPoint.point.distanceTo(mousePos) <= this.radius / 400)
      
      if (canverControlFirstMatch >= 0) {
        this.selected = {
          isTop: false,
          index: canverControlFirstMatch
        };
        return;
      }		
      
			return;
    },
    mouseup: function (): void {
      this.isDrag = false;
      this.selected = null;
    },
    mousemove: function (event: {target: HTMLCanvasElement; clientX: number; clientY: number }): void {
      
			if (this.selected === null || this.isDrag === false) {
				return;
      }
      
			const rect = event.target.getBoundingClientRect();
      const mousePos = new Vector2(event.clientX - rect.left - 100, -(event.clientY - rect.top - 100)).divideScalar(400);
      if (this.selected.isTop) {
        this.topControls.points[this.selected.index].point = mousePos.clone();
      } 
      else {
        this.canverControls.points[this.selected.index].point = mousePos.clone();
      }

      this.draw();
		},
    draw: function (): void {
      this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height)
			this.drawCircle();
			this.drawLine();
    },
    drawCircle: function (): void {
      const curves = [this.topControls, this.canverControls];
      curves.forEach((curve) => {
        curve.points
          .filter((item) => item.visible)
          .map((item) => item.point)
          .forEach((point) => {
            this.ctx.beginPath();
            this.ctx.arc(
              point.x * 400 + 100,
              -point.y * 400 + 100,
              this.radius,
              0,
              Math.PI * 2,
              false
            );
            this.ctx.fillStyle = curve.controlPointStyle;
            this.ctx.fill();
            this.ctx.strokeStyle = curve.controlPointStyle;
            this.ctx.stroke();
          });
      });
    },

    drawLine(): void {
      this.ctx.beginPath();
      this.ctx.moveTo(500, 100);
      this.getShape.forEach((point) => {
        this.ctx.lineTo(point.x * 400 + 100, -point.y * 400 + 100);
			});
			this.ctx.strokeStyle = 'red';
      this.ctx.closePath();
      this.ctx.stroke();
    },
  },
});
</script>