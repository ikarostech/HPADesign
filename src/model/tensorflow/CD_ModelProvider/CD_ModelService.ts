import * as tf from '@tensorflow/tfjs';

import { AIAirfoilPoint } from '../AIAirfoilPoint/AIAirfoilPoint';

// eslint-disable-next-line @typescript-eslint/class-name-casing
export class CD_ModelService {

  private constructor(model: tf.LayersModel) {
    this.model = model
  }

  private model: tf.LayersModel;
  public static async newBuild(): Promise<CD_ModelService> {
    const model = await tf.loadLayersModel("https://raw.githubusercontent.com/ikarostech/DW/CD/model.json");
    const service = new CD_ModelService(model);
    return service;
  }

  public predict(point: AIAirfoilPoint): number {
    return Number(this.model.predict(point.getTFTensor()).toString().replace(/[^0-9|.]/g, '')) / 100;
  }

}