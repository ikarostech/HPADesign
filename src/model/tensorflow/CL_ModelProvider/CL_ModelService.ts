import * as tf from '@tensorflow/tfjs';

import { AIAirfoilPoint } from '../AIAirfoilPoint/AIAirfoilPoint';

export class CL_ModelService {

  private constructor(model : tf.LayersModel) {
    this.model = model
  }

  private model: tf.LayersModel;
  public static async newBuild() : Promise<CL_ModelService> {
    let model = await tf.loadLayersModel("https://raw.githubusercontent.com/ikarostech/DW/master/model.json");
    let service = new CL_ModelService(model);
    return service;
  }

  public predict(point : AIAirfoilPoint): number {
    return Number(this.model.predict(point.getTFTensor()).toString().replace(/[^0-9|.]/g, ''));
  }
  
}