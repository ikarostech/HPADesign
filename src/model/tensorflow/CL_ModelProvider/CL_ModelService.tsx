import * as tf from '@tensorflow/tfjs';

import AIAirfoilPoint from '../AIAifoilPoint/AIAirfoilPoint';

export class CL_ModelService {

  private constructor(model : tf.LayersModel) {
    this.model = model
  }

  private model: tf.LayersModel;
  public static async newBuild() : Promise<CL_ModelService> {
    let model = await tf.loadLayersModel("https://raw.githubusercontent.com/ikarostech/DW/master/model.json");
    let service = new CL_ModelService(model);
    return service;
  }

  public predict(point : AIAirfoilPoint): Number {
    return Number(this.model.predict(point.getTFShape()).toString().replace(/[^0-9|.]/g, ''));
  }
  
}