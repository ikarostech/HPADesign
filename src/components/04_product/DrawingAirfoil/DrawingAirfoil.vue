<template>
  <div>
    <BezierFoilCanvas 
      @shapeUpdate = "shapeUpdate"
      :radius="10" 
    />
    <AIAirfoilCanvas/>
  </div>
</template>
<script lang="ts">
import Vue from 'vue';
import BezierFoilCanvas from '@/components/01_element/DrugableCanvas/BezierFoilCanvas.vue'
import AIAirfoilCanvas from '@/components/03_composition/AIAirfoilCanvas/AIAirfoilCanvas.vue';

import { CanvasPoint } from '@/components/01_element/DrugableCanvas/CanvasPoint';
import { BezierControls } from '@/components/01_element/DrugableCanvas/BezierControls';
import { Vector2 } from 'three/src/math/Vector2';


export default Vue.extend({
  name: 'App',
  components: {
    BezierFoilCanvas,
    AIAirfoilCanvas
  }
});
</script>