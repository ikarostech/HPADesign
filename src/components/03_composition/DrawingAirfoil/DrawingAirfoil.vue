<template>
  <div>
    <BezierFoilCanvas @updateShape="updateShape" :radius="5" />
    <airfoil-export-button :shape="shape"></airfoil-export-button>
  </div>
</template>
<script lang="ts">
import Vue from "vue";
import BezierFoilCanvas from "@/components/01_element/DrugableCanvas/BezierFoilCanvas.vue";
import AirfoilExportButton from "@/components/01_element/ExportButton/Airfoil/AirfoilExportButton.vue";
import airfoil from "@/model/component/airfoil/Airfoil"

import { Vector2 } from "three/src/math/Vector2";
import { CL_ModelService } from "@/model/tensorflow/CL_ModelProvider/CL_ModelService";
import { CD_ModelService } from "@/model/tensorflow/CD_ModelProvider/CD_ModelService";
import { AIAirfoilPoint } from "@/model/tensorflow/AIAirfoilPoint/AIAirfoilPoint";

export type DataType = {
  shape: Vector2[];
  airfoil: airfoil | null;
  cl_service: CL_ModelService | null;
  cd_service: CD_ModelService | null;
};

export default Vue.extend({
  name: "App",
  components: {
    BezierFoilCanvas,
    AirfoilExportButton,
  },
  data: function (): DataType {
    return {
      shape: [],
      airfoil: null,
      cl_service: null,
      cd_service: null,
    };
  },
  methods: {
    updateShape: function (shape: Vector2[]): void {
      this.shape = shape;
    },
  },
  watch: {
    shape: function (val: Vector2[]): void {
      if (
        this.shape.length === 0 ||
        this.cl_service === null ||
        this.cd_service === null
      ) {
        this.airfoil = null
        return;
      }
      
      const airfoilPoint: AIAirfoilPoint = new AIAirfoilPoint(val)
      const C_L = this.cl_service.predict(airfoilPoint);
      const C_D = this.cd_service.predict(airfoilPoint);

      this.airfoil = new airfoil(C_L, C_D)

      this.$store.dispatch("updateAirfoil",this.airfoil);
    },
  },
  mounted: async function () {
    this.cl_service = await CL_ModelService.newBuild();
    this.cd_service = await CD_ModelService.newBuild();
  },
});
</script>