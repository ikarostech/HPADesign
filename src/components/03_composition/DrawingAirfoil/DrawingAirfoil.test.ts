import 'jest';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex, { Store } from 'vuex';
import BezierFoilCanvas from '@/components/01_element/DrugableCanvas/BezierFoilCanvas.vue';
import DrawingAirfoil from './DrawingAirfoil.vue';
import { CL_ModelService } from '@/model/tensorflow/CL_ModelProvider/CL_ModelService';
import { CD_ModelService } from '@/model/tensorflow/CD_ModelProvider/CD_ModelService';
import { Vector2 } from 'three/src/math/Vector2';


describe('DrawingAirfoil', () => {
  let actions: { updateC_L: () => void; updateC_D: () => void }
  let store: Store<{}>

  const localVue = createLocalVue();
  localVue.use(Vuex);

  const wrapper = shallowMount(DrawingAirfoil, {});
  const component = (wrapper.vm as any);

  beforeEach(() => {
    actions = {
      updateC_L: jest.fn(),
      updateC_D: jest.fn(),
    }

    store = new Vuex.Store({
      state: { C_L: 0, C_D: 0 },
      actions
    })
  })

  it('snapshot test', () => {
    // when&then
    expect(wrapper.element).toMatchSnapshot();
  });

  it('shape正常系 - emit:updatedShapeを受け取る', () => {
    // when
    const shape: Vector2[] = [new Vector2(1, 1)];

    // then
    wrapper.findComponent(BezierFoilCanvas).vm.$emit('updateShape', shape);
    expect((wrapper.emitted().updateShape as Array<any>)[0][0]).toBe(shape);
  });
  it('shape正常系 - emitで受け取ったデータを自分のdataに入れる', () => {
    // when
    const shape: Vector2[] = [new Vector2(1, 1)];

    // then
    component.updateShape(shape);
    expect(component.shape).toBe(shape);
  })
  it("C_L更新時にstore/C_Lの値を更新", async () => {
    //given
    const wrapper = shallowMount(DrawingAirfoil, { store, localVue });
    const component = (wrapper.vm as any);
    // そのまま使うとundefinedになってしまうため無理やり型を入れてしまう
    component.cl_service = await CL_ModelService.newBuild();
    component.cd_service = await CD_ModelService.newBuild();
    jest.spyOn(component.cl_service, 'predict').mockReturnValue(-1);
    jest.spyOn(component.cd_service, 'predict').mockReturnValue(-1);

    // when
    component.updateShape([new Vector2(1, 1)]);

    // then
    wrapper.vm.$nextTick(() => {
      expect(component.shape).toEqual([new Vector2(1, 1)]);
      console.log(component.C_L);
      expect(component.C_L).toEqual(-1);
      expect(actions.updateC_L).toBeCalled()
    })
  });

  it("C_D更新時にstore/C_Dの値を更新", async () => {
    // given
    const wrapper = shallowMount(DrawingAirfoil, { store, localVue });
    const component = (wrapper.vm as any);
    // そのまま使うとundefinedになってしまうため無理やり型を入れてしまう
    component.cl_service = await CL_ModelService.newBuild();
    component.cd_service = await CD_ModelService.newBuild();
    jest.spyOn(component.cl_service, 'predict').mockReturnValue(-1);
    jest.spyOn(component.cd_service, 'predict').mockReturnValue(-1);

    // when
    component.updateShape([new Vector2(1, 1)]);

    // then
    wrapper.vm.$nextTick(() => {
      expect(component.shape).toEqual([new Vector2(1, 1)]);
      console.log(component.C_D);
      expect(component.C_D).toEqual(-1);
      expect(actions.updateC_D).toBeCalled()
    })


  });

  it('C_L異常系 - shapeが空の時にC_Lの値を0にする', async () => {
    // given
    // そのまま使うとundefinedになってしまうため無理やり型を入れてしまう
    component.cl_service = await CL_ModelService.newBuild();

    // when
    component.updateShape([]);

    // then
    expect(component.shape).toEqual([]);
    expect(component.C_L).toEqual(0);
  });

  it('C_L異常系 - shapeが空の時にC_Dの値を0にする', async () => {
    // given
    // そのまま使うとundefinedになってしまうため無理やり型を入れてしまう
    component.cl_service = await CD_ModelService.newBuild();

    // when
    component.updateShape([]);

    // then
    expect(component.shape).toEqual([]);
    expect(component.C_D).toEqual(0);
  });

  afterEach(() => {
    jest.clearAllMocks();
  })
});