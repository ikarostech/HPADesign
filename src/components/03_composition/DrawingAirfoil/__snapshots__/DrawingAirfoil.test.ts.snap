// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawingAirfoil snapshot test 1`] = `
<div>
  <bezierfoilcanvas-stub
    radius="5"
  />
   
  <p>
    C_L = 0
  </p>
   
  <p>
    C_D = 0
  </p>
   
  <airfoil-export-button-stub
    shape=""
  />
</div>
`;
